cppcache/src/TcrEndpoint.cpp

```diff
@@ -1155,6 +1157,10 @@ void TcrEndpoint::setConnectionStatus(bool status) {
       m_isActiveEndpoint = false;
       LOGFINE("Disconnected from endpoint %s", m_name.c_str());
       triggerRedundancyThread();
+      // Test hook
+      if (TcrEndpoint::TEST_DISCONNECTIONS) {
+        TcrEndpoint::listOfDisconnectedServers.push_back(m_name.c_str());
```

Comment:
I am strongly against adding any static methods and variables without a really strong argument for why they can't be member instance variables.

I think it would make for a cleaner solution to and an API to the owner of `TcrEndpoint` collections, I think this might be `CacheImpl`, that allows for the registration of an event handler. Then on any of these `TcrEndpoint` status changes we would invoke the registered handlers. The test could then register handler it needs and asserts that it is invoked, without the need for global variables.

Reply:
Thanks for comments!

Would it be acceptable to use existing counter and method `numberOfTimesFailed()` from `TcrEndpoint` to check for disconnections?

```
int32_t numberOfTimesFailed() { return m_numberOfTimesFailed; }
 ...
void TcrEndpoint::setConnectionStatus(bool status) {
  std::lock_guard<decltype(m_connectionLock)> guard(m_connectionLock);
    if (m_connected != status) {
      bool connected = m_connected;
      m_connected = status;
      if (connected) {
        m_numberOfTimesFailed += 1;
        m_isAuthenticated = false;
        // disconnected
        LOGFINE("Disconnecting from endpoint %s", m_name.c_str());
        closeConnections();
        ...
```
I think it would be possible to use it, because `TcrEndpoint` object is not destroyed when endpoint disconnects from client.

It would be used in test case something like this:
```
auto cacheImpl = CacheRegionHelper::getCacheImpl(cache.get());

// shut down one endpoint (server)

ASSERT_EQ(cacheImpl->getNumberOfTimeEndpointDisconnected(poolName, epRunningHostname), 0);
ASSERT_GT(cacheImpl->getNumberOfTimeEndpointDisconnected(poolName, epDisconnectedHostname), 0);

```
`ThinClientPoolDM` is actually owner of `TcrEndpoint` collection, so `getNumberOfTimeEndpointDisconnected(endpoint)` would basically find given endpoint and return it's `numberOfTimesFailed()`.